Enables use of single/dual channel textures

sarbes opened the pull request:
## Description
This PR enables skinners to use single/dual channel textures by naming them "filename_[channelusage].png".

[channelusage] can be the following values:
- "alpha" is a single channel texture, which is equivalent to [FF,FF,FF,channel1]
- "luma" is a single channel texture, which is equivalent to [channel1,channel1,channel1,FF]
- "lumaalpha" is a dual channel texture, which is equivalent to [channel1,channel1,channel1,channel2]

## Motivation and context
Currently, GPU textures in Kodi are often (always?) stored as BGRA, irregardless of the actual content. This wastes a lot of resources, as a lot of assets are only greyscale. For Estuary, this is the case for >90% of all assets. 

Replacing the colored textures would save a large amount of memory bandwidth. The background textures of Estuary are ~10MB large. This occupies >600MB/s memory bandwidth at 1080p60(might be lower with compression), which is a lot on low end systems. Single channel textures would save 75% of the BW.

Also, compression (disk memory) might be a bit better on when using single channel textures.

For OpenGL >=3.3 (or suitable extensions) and OpenGL ES 2.0, "luma", "alpha" and lumaalpha hints are supported. In every other case, the texture gets converted into a full GBRA texture.

## How has this been tested?
Compiles and runs fine with GL and GLES.

DX is a shot in the blue, but fortunately it is not complex.

## What is the effect on users?
Moar speed on lower end systems.

## Screenshots (if appropriate):
Here is what Renderdoc reported before the change:
![before](https://user-images.githubusercontent.com/30039775/173712473-9ece6a21-ad18-4336-beac-6ca8bc3db422.png)

And after the change for GL:
![after_gl](https://user-images.githubusercontent.com/30039775/173712288-ba9ce0d9-2334-4e27-9328-6e14041b3d22.png)

...and for GLES 2.0:
![after_gles](https://user-images.githubusercontent.com/30039775/173712292-9d9e324a-13f8-4239-b53f-c78e5dc7a7e3.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

garbear commented on xbmc/guilib/Texture.cpp:
```diff
@@ -493,3 +497,43 @@ bool CTexture::IsMipmapped() const
+  unsigned char* newPixels;
+
+  size_t size = GetPitch() * GetRows();
+  newPixels = static_cast<unsigned char*>(KODI::MEMORY::AlignedMalloc(size * 4, 32));
```
I don't like uninitialized variables, because the memory is invalid between the lines where it's declared and defined, so instead we could do:

```c++
size_t size = GetPitch() * GetRows();

unsigned char* oldPixels = m_pixels;
unsigned char* newPixels = static_cast<unsigned char*>(KODI::MEMORY::AlignedMalloc(size * 4, 32));
```

sarbes replied:
Done.

I hope I remember this in the future.

garbear commented on xbmc/guilib/Texture.cpp:
```diff
@@ -493,3 +497,43 @@ bool CTexture::IsMipmapped() const
+  newPixels = static_cast<unsigned char*>(KODI::MEMORY::AlignedMalloc(size * 4, 32));
+
+  if (newPixels == nullptr)
+    CLog::Log(LOGERROR, "{} - Could not allocate {} bytes. Out of memory.", __FUNCTION__, size * 4);
```
Should we return early here? It looks like `newPixels` is used later.

sarbes replied:
Yeah.

It might look differently on GL/GLES/DX if this happens, but I guess this would be the least of the problems.

garbear reviewed (approved):
Cool feature, cleanly added. And great PR description.

What's the behavior on DX and GL systems that don't support the feature? Or would the feature be considered an initial attempt with all platforms being covered later?

sarbes commented:
On unsupported systems, the texture gets converted via ConvertToBGRA(). For older GL systems without swizzling, I would need to add a dozen or so shaders (I'm certain Lukas would rip my head off if I would do that :D). And I'm staying away from DX. From what I've gathered, there is no swizzling, so everything would have to be done in shaders.

I'll have to add some defines for systems without newer GL headers (i.e. TVOS). Together with your suggestions.

sarbes commented:
Sorry @garbear, I was not happy with the previous approach. I've rewrote the selection logic behind it.

Now, the file extension is used to determine the texture property. This is a better approach as it will avoid situations where the skin might want to have different versions of the textures. Because it is not a skin attribute anymore, there is no need to transition all assets as it might be the case before. I can also integrate it into the texture packer in the future. A file is identified by naming it "texturename.luma.png".

Also on the plus side, we do have more options. Besides the "alpha" and "luma" qualifier, there is also "intensity" and "lumaalpha" The latter is a dual channel texture.

New for GLES 2.0 is the option to use dual channel textures if the extension "GL_EXT_texture_rg" is available. In this case,  "alpha" and "intensity" will be converted to a dual channel texture, saving 50% resources.

sarbes commented:
GL_GREEN is missing for iOS, I'll add the define when Jenkins is done.

garbear commented on xbmc/guilib/TextureManager.cpp:
```diff
@@ -297,6 +298,24 @@ const CTextureArray& CGUITextureManager::Load(const std::string& strTextureName,
+  unsigned int textureFormat = XB_FMT_A8R8G8B8;
+  if (StringUtils::StartsWith(strTextureName, "special://skin/"))
+  {
+    std::vector<std::string> s = StringUtils::Split(strTextureName, ".");
```
What do we pay you by, the character? :stuck_out_tongue_winking_eye: No one ever underappreciates a good explanatory variable name.

sarbes replied:
Done.

garbear commented on xbmc/guilib/Texture.cpp:
```diff
@@ -493,3 +507,117 @@ bool CTexture::IsMipmapped() const
+  if (newPixels == nullptr)
+  {
+    CLog::Log(LOGERROR, "{} - Could not allocate {} bytes. Out of memory.", __FUNCTION__, newSize);
+    m_format = oldFormat;
```
Avoid effects in early returns, as all future early returns would need to duplicate the effect. I would only update internal state on success.

sarbes replied:
This changed with the new commit.

garbear commented:
> I've renamed the original format XB_FMT_A8 to XB_FMT_R8, as this reflects its usage far more accurate.

Can you explain this as if I don't know what any of those letters are abbreviations for?

garbear commented:
I think I see an area of improvement. We can currently convert between texture formats on the fly, back and forth. It would improve performance to not free old pixels and then having to regenerate them.

If freeing old pixels is still desired, then I would argue to move the mutation out of the texture. Let the calling object create a new texture of a different format and free the old texture, moving old/new memory management out of CTexture.

sarbes commented:
> > I've renamed the original format XB_FMT_A8 to XB_FMT_R8, as this reflects its usage far more accurate.
> 
> Can you explain this as if I don't know what any of those letters are abbreviations for?

Previously, XB_FMT_A8 stand for a generic single channel texture. On OpenGL and DX, it was used as the red component for font. I've changed it to reflect this aspect. I would say it stands for XBMC_FORMAT_APLHA8{bit].

While it is possible to use the identifier for other kinds of texture representations, the actual intent of what the texture is has to be stored somewhere. So the new identifiers were born, reflecting what the texture is representing.


> I think I see an area of improvement. We can currently convert between texture formats on the fly, back and forth. It would improve performance to not free old pixels and then having to regenerate them.
> 
> If freeing old pixels is still desired, then I would argue to move the mutation out of the texture. Let the calling object create a new texture of a different format and free the old texture, moving old/new memory management out of CTexture.

I could allocate memory for a whole four channel texture all the time and work in this space without an extra alloc. Is that what you mean?

lrusak commented on xbmc/guilib/Texture.cpp:
```diff
@@ -217,7 +227,7 @@ std::unique_ptr<CTexture> CTexture::LoadFromFile(const std::string& texturePath,
   }
 #endif
-  std::unique_ptr<CTexture> texture = CTexture::CreateTexture();
+  std::unique_ptr<CTexture> texture = CTexture::CreateTexture(0, 0, format);
```
Is there any use for the default params in `CreateTexture` anymore?

sarbes replied:
Should I remove them?

lrusak commented on xbmc/guilib/TextureManager.cpp:
```diff
@@ -297,6 +298,24 @@ const CTextureArray& CGUITextureManager::Load(const std::string& strTextureName,
+    std::vector<std::string> strTextureNameSplitted = StringUtils::Split(strTextureName, ".");
+    if (strTextureNameSplitted.size() >= 3)
+    {
+      std::string possibleFormat = strTextureNameSplitted[strTextureNameSplitted.size() - 2];
```
What is the meaning of `-2` here? is there a better way to find what you are looking for in the list?

sarbes replied:
Done.

lrusak commented on xbmc/guilib/Texture.cpp:
```diff
@@ -55,6 +56,7 @@ void CTexture::Allocate(unsigned int width, unsigned int height, unsigned int fo
   m_imageWidth = m_originalWidth = width;
   m_imageHeight = m_originalHeight = height;
   m_format = format;
+  unsigned int allocationFormat = format;
```
I would rather see a sized type (`uint32_t`) for new code but no sweat here really.

sarbes replied:
1. I've made the selection logic a bit simpler.
2. Done.

lrusak commented on xbmc/guilib/TextureGL.cpp:
```diff
@@ -111,6 +141,35 @@ void CGLTexture::LoadToGPU()
   }
 
 #ifndef HAS_GLES
+  if (CServiceBroker::GetRenderSystem()->IsExtSupported("GL_ARB_texture_swizzle"))
```
might also need to check for `GL_EXT_texture_swizzle`

sarbes replied:
I've made it a bit simpler.

lrusak commented on xbmc/guilib/TextureGL.cpp:
```diff
@@ -18,6 +19,35 @@
+#define GL_TEXTURE_SWIZZLE_RGBA 0x8E46
+#endif
+#ifndef GL_RG_EXT
+#define GL_RG_EXT 0x8E46
```
This value seems wrong?
```
$ grep -R -E "GL_RG(_EXT)? " /usr/include                                                                                                                                                  
/usr/include/GL/glext.h:#define GL_RG                             0x8227
/usr/include/GL/glcorearb.h:#define GL_RG                             0x8227
/usr/include/GLES2/gl2ext.h:#define GL_RG_EXT                         0x8227
/usr/include/GLES3/gl31.h:#define GL_RG                             0x8227
/usr/include/GLES3/gl3.h:#define GL_RG                             0x8227
/usr/include/GLES3/gl32.h:#define GL_RG                             0x8227
```

sarbes replied:
Yes, it is. It was a bit late when I copied it. Also, I "GL_RG" would be more appropriate

But really, it should not matter as the code path containing it should never been taken on platforms not defining it. Should I check for that on compile time?

lrusak reviewed (commented):
Looks nice! Cool feature! I left a few comments.

I think there may be issues with GLES 2.0 as the mixture of defines and whatnot may lead to unexpected results. It would be good to test on GLES 2.0 only hardware if possible.

sarbes commented:
Huh, seems to me that Mali-450 GPU's do support GL_INTENSITY and swizzling in hardware. But GLES 2.0 does not expose this function, and there is no suitable extension. What a mess.

Lima supports those with OpenGL 2.1.

fritsch commented:
Maybe time for a state-of-the OpenGLES 3.1+ path?

sarbes commented:
Na, Not for this. I'm totally fine writing the fallback for this minor instance. It is not restrictive.

sarbes commented on xbmc/guilib/TextureManager.cpp:
```diff
@@ -297,6 +298,24 @@ const CTextureArray& CGUITextureManager::Load(const std::string& strTextureName,
+      std::string possibleFormat = strTextureNameSplitted[strTextureNameSplitted.size() - 2];
+      if (StringUtils::EqualsNoCase(possibleFormat, "luma"))
+        textureFormat = XB_FMT_L8;
+      if (StringUtils::EqualsNoCase(possibleFormat, "alpha"))
```
Done.

sarbes commented on xbmc/guilib/TextureManager.cpp:
```diff
@@ -297,6 +298,24 @@ const CTextureArray& CGUITextureManager::Load(const std::string& strTextureName,
+        textureFormat = XB_FMT_L8;
+      if (StringUtils::EqualsNoCase(possibleFormat, "alpha"))
+        textureFormat = XB_FMT_A8;
+      if (StringUtils::EqualsNoCase(possibleFormat, "intensity"))
```
Done.

sarbes commented on xbmc/guilib/TextureManager.cpp:
```diff
@@ -297,6 +298,24 @@ const CTextureArray& CGUITextureManager::Load(const std::string& strTextureName,
+        textureFormat = XB_FMT_A8;
+      if (StringUtils::EqualsNoCase(possibleFormat, "intensity"))
+        textureFormat = XB_FMT_I8;
+      if (StringUtils::EqualsNoCase(possibleFormat, "lumaalpha"))
```
Done.

sarbes commented on xbmc/guilib/TextureManager.cpp:
```diff
@@ -297,6 +298,24 @@ const CTextureArray& CGUITextureManager::Load(const std::string& strTextureName,
+  unsigned int textureFormat = XB_FMT_A8R8G8B8;
+  if (StringUtils::StartsWith(strTextureName, "special://skin/"))
+  {
+    std::vector<std::string> strTextureNameSplitted = StringUtils::Split(strTextureName, ".");
```
Done.

sarbes commented:
I've rewritten some parts and pushed a new version. The highlights:
- GL_INTENSITY type textures have been dropped. I don't see much use.
- OpenGL ES 2.0 supports luma and luma/alpha textures "normally", and pure alpha variants via shaders.
- OpenGL >=3.3 supports all via swizzling.
- All other platforms decode the texture to full BGRA.
- The texture packer is now able to pack such textures from PNG files. Extensions like "filename.luma.png" will be shortened to "filename.png", so that the hard coded default textures can be supplied.

Now, I have to figure out how to resolve the conflict (:.

sarbes commented on tools/depends/native/TexturePacker/src/decoder/PNGDecoder.cpp:
```diff
@@ -140,14 +141,19 @@ bool PNGDecoder::LoadFile(const std::string &filename, DecodedFrames &frames)
   png_get_IHDR(png_ptr, info_ptr, &temp_width, &temp_height, &bit_depth, &color_type,
                NULL, NULL, NULL);
 
+  //if (color_type != PNG_COLOR_TYPE_GRAY &&
```
I'll clean this up.

sarbes commented on tools/depends/native/TexturePacker/src/decoder/PNGDecoder.cpp:
```diff
@@ -217,12 +231,31 @@ bool PNGDecoder::LoadFile(const std::string &filename, DecodedFrames &frames)
+    frame.rgbaImage.bbp = 8;
+    frame.rgbaImage.pitch = temp_width;
+  }
+  else if (color_type == XB_FMT_L8A8)
```
Sorry about that. It's getting late :/

sarbes commented on tools/depends/native/TexturePacker/src/XBTFWriter.cpp:
```diff
@@ -18,6 +18,7 @@
  *
  */
 
+#include <cstdint>
```
My IDE thinks it is smart. I'll fix that.

garbear reviewed (approved):
I gave the most recent state of the PR a full read through. I don't see any issues left so I'm leaving my +1.

garbear commented:
jenkins build this please

lrusak commented:
It would be nice if you could squash the last two commits into the relevant prior commits.

This is probably easiest to reset the commit, do a partial add, then commit as a fix up, and autosquash in a rebase.

Something like
```
git reset HEAD^

git add -p
git commit --fixup=<sha-of-relevant-commit>

Repeat last two commands till all changes are committed.

git rebase -i --autosquash upstream/master
```

sarbes committed:
Renaming of the texture format A8 to R8

sarbes committed:
Add GLES shaders for single/dual channel textures

sarbes commented on xbmc/guilib/TextureGL.cpp:
```diff
@@ -224,3 +296,45 @@ void CGLTexture::BindToUnit(unsigned int unit)
+      return false;
+  }
+#endif
+}
```
Dang it.

sarbes committed:
Implement single/dual channel textures for GL/GLES

ilmich commented:
Hi @sarbes 

from what I understand, to make the change effective you need to apply changes to the skin (e.g. estuary).
It's correct?! 
If yes, the rule for renaming files can be
- jpg --> luma
- png b/w --> alpha
- others --> lumaalpha

or it always depends on the skin and not on the file (if it's in color or not)

ilmich commented:
Hi @sarbes 

building kodi (nexus 20.1) with this patch and [#22919](https://github.com/xbmc/xbmc/pull/22919) I can say that the fluidity of my rk3229/ddr2 box has improved a lot and I haven't noticed any problems of any kind.

Thank you very much

sarbes commented:
In this iteration, the packer is looking at the format of the PNG and for special "markets" in the file name. This needs proper file handling on the skinner side (by providing grescale textures).

But I've redone the texture packer again. Now, it scans through all the pixels, and chooses a proper format based on its findings. With it, even a full RGBA PNG will be stored as an alpha only texture, if the color channels are all 0xff. No skin/texture changes needed, provided that textures are properly done. I'll commit it soon.

In practice, this means that Estuary reduces its texture footprint by over 60%, as it is composed mainly out of alpha only textures. Some textures (around 10%) are broken (having varying colour information) and are stored as luma/alpha textures, which has to be fixed separately.

For Estuary, the most performance uplift with this patch would come from providing greyscale background textures. Either replace the JPGs in skin.estuary/extras/backgrounds/ with greyscale versions (I have not found a way to strip the chroma channels without reencoding) or attach the "_luma.jpg" hint (+update the XML). At 60fps, this saves over 540MB/s.

sarbes committed:
Implement L/A/LA support for the texture packer

sarbes commented:
The texture packer now reduces textures to a suitable format.

ilmich commented:
Hi @sarbes,

I did some tests renaming the files but didn't notice the expected improvements. Surely I'm doing something wrong, or the jpegs should be treated differently.
But you hit the 'problem' because by mistake in one of my tests, I completely removed the background(typo in filename) and estuary became ultra fast.

sarbes commented:
@ilmich I've double checked it with Renderdoc. Skin (XBT) textures are uploaded as GL_LUMINANCE/GL_ALPHA​/GL_LUMINANCE_ALPHA​ on GLES. 

I forgot to update my top post. Textures should be marked with an underscore (e.g. "primary_luma.jpg"). This might be your problem.

Another way to get "automatic" texture format handling is by providing compatible image formats. Greyscale JPEGs (without any chroma channels) will decode to GL_LUMINANCE textures. Greyscale PNGs will decode to GL_LUMINANCE or GL_LUMINANCE_ALPHA​, depending on the present of an alpha channel.

ilmich commented:
Hi @sarbes,

for my tests with libreelec I used this snippet for jpeg images
```
for file in ${INSTALL}/usr/share/kodi/addons/skin.estuary/extras/backgrounds/*.jpg; do 
        mogrify "$file" -colorspace Gray -quality 90
done
```
which should allow the use of luma textures.
I don't know if I did everything correctly, I see a slight improvement. But keep in mind that I use them with Nexus (they still apply without problems) and that I probably approached hardware limitations, rather than software inefficiencies.

sarbes commented:
> which should allow the use of luma textures. I don't know if I did everything correctly, I see a slight improvement. But keep in mind that I use them with Nexus (they still apply without problems) and that I probably approached hardware limitations, rather than software inefficiencies.

This PR should relief the memory bus quite a bit. There are three possibilities:
- Your device is not memory bandwidth limited
- Your device is GPU bound (maybe)
- The driver is converting the texture into an RGBA texture internally (shouldn't be)

You can test if your device is memory bandwidth limited by replacing those textures with 1x1 pixel textures. If the framerate is increasing, you are bandwidth bound.

yol commented on xbmc/guilib/FFmpegImage.cpp:
```diff
@@ -349,12 +356,43 @@ void CFFmpegImage::FreeIOCtx(AVIOContext** ioctx)
+    outputPixelFormat = AV_PIX_FMT_RGB32;
+  else
+  {
+    CLog::Log(LOGERROR, "{} - No valid texture format ({}) passed", __FUNCTION__, textureFormat);
```
```suggestion
    CLog::LogF(LOGERROR, "No valid texture format ({}) passed", textureFormat);
```

yol commented on xbmc/guilib/Texture.cpp:
```diff
@@ -493,3 +536,41 @@ bool CTexture::IsMipmapped() const
+{
+  size_t size = GetPitch(m_textureWidth, format) * GetRows(m_textureHeight, format);
+
+  if (format == XB_FMT_A8)
```
could be a switch/case

yol commented on xbmc/guilib/Texture.cpp:
```diff
@@ -493,3 +536,41 @@ bool CTexture::IsMipmapped() const
+    for (int32_t i = size - 1; i >= 0; i--)
+    {
+      m_pixels[i * 4 + 3] = m_pixels[i];
+      m_pixels[i * 4 + 2] = char(0xff);
```
so why all the `char` casts? m_pixels is unsigned char?

yol commented on xbmc/guilib/TextureDX.h:
```diff
@@ -24,6 +24,7 @@ class CDXTexture : public CTexture
   void DestroyTextureObject();
   virtual void LoadToGPU();
   void BindToUnit(unsigned int unit);
+  bool IsGPUFormatSupported(const uint32_t format) override;
```
```suggestion
  bool IsGPUFormatSupported(uint32_t format) override;
```

yol commented on xbmc/rendering/gles/RenderSystemGLES.cpp:
```diff
@@ -443,6 +452,36 @@ void CRenderSystemGLES::InitialiseShaders()
     CLog::Log(LOGERROR, "GUI Shader gles_shader_multi_blendcolor.frag - compile and link failed");
   }
 
+  m_pShader[ShaderMethodGLES::SM_MULTI_BLENDCOLOR_COLOR_ALPHA] =
```
all of this c&p could really use a refactor

ilmich commented:
Hi @sarbes 

I have tested this patch daily and have not seen any bugs.

> This PR should relief the memory bus quite a bit. There are three possibilities:
> 
> * Your device is not memory bandwidth limited
> * Your device is GPU bound (maybe)
> * The driver is converting the texture into an RGBA texture internally (shouldn't be)
> 
> You can test if your device is memory bandwidth limited by replacing those textures with 1x1 pixel textures. If the framerate is increasing, you are bandwidth bound.

I wrote wrong,

- without your patch, my device has memory bandwidth limits
- with your patch the memory limits disappear and therefore I get limits on the gpu
- converting the background jpeg does not change anything having reached the limits of the GPU

However, it's an excellent result and I thank you for your work.

CrystalP commented on system/shaders/GLES/2.0/gles_shader_multi_blendcolor_alpha_alpha.frag:
```diff
@@ -0,0 +1,44 @@
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
```
out of curiosity, is the whole thing necessary?
Some places just have
/*
 *  Copyright (C) 2023 Team Kodi
 *  This file is part of Kodi - https://kodi.tv
 *
 *  SPDX-License-Identifier: GPL-2.0-or-later
 *  See LICENSES/README.md for more information.
 */

garbear replied:
Probably because this file was copied from an existing file with our old header. It should be updated.

CrystalP commented:
Can you please rebase. I want to look at the D3D part but there have been quite a few build changes for Windows recently and switching backwards and forwards is painful.

CrystalP commented on xbmc/guilib/Texture.h:
```diff
@@ -106,12 +109,14 @@ class CTexture
   static unsigned int PadPow2(unsigned int x);
   static bool SwapBlueRed(unsigned char *pixels, unsigned int height, unsigned int pitch, unsigned int elements = 4, unsigned int offset=0);
 
+  virtual bool IsGPUFormatSupported(const uint32_t format) { return format == XB_FMT_A8R8G8B8; }
```
const

sarbes commented:
The PR has been (partly) superceded by the one linked above. 

Thanks to everyone involved, and sorry for the inconvenience.
